### PR TITLE
Issue a warning on restore about obsolete DotNetCliToolReference

### LINKED ISSUE #2014: Add obsolete warning for DotNetCliToolReference packages that are now built-in commands
A few ASP.NET Core packages that used to be installed via DotNetCliToolReference will be built-in commands for the CLI.  When invoking "dotnet watch" now, it will use the built-in version, not the one delivered via DotNetCliToolReference, so DotNetCliToolReference should be removed.
cref https://github.com/aspnet/Announcements/issues/290.

Since we don't have a proper way to issue obsolete messages about packages (NuGet/Home#2867), can we add a target that runs before Restore that issues an obsolete warning for the following packages

```xml
<ItemGroup>
   <DotNetCliToolReference Include="Microsoft.DotNet.Watcher.Tools" Version="(all versions)" />
   <DotNetCliToolReference Include="Microsoft.Extensions.SecretManager.Tools" Version="(all versions)" />
   <DotNetCliToolReference Include="Microsoft.Extensions.Caching.SqlConfig.Tools" Version="(all versions)" />
</ItemGroup>
```

Proposed text:

> warning SDK0000: the DotNetCliToolReference to Microsoft.DotNet.Watcher.Tools is obsolete. This tool is now bundled in the CLI. The DotNetCliToolReference can be removed.

### PR DESCRIPTION
**Pull request by natemcmaster**
Resolves #2014 

I will make a corresponding change to dotnet/cli to generate Microsoft.NETCoreSdk.BundledCliTools.props with a list of obsolete package IDs to match https://github.com/dotnet/cli/blob/master/build/BundledDotnetTools.props.

VS:

image

Command line:

image

FYI @muratg @DamianEdwards @prafullbhosale @divega @bricelam @Eilon

### CODE REVIEW

**Review by livarcocc — APPROVED** [review]

File: src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.ObsoleteReferences.targets [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+    <_ReferenceToObsoleteDotNetCliTool Include="@(DotNetCliToolReference)" />
+    <DotNetCliToolReference Remove="@(BundledDotNetCliToolReference)" />
+    <_ReferenceToObsoleteDotNetCliTool Remove="@(DotNetCliToolReference)" />
+  </ItemGroup>
+
+  <Target Name="_WarnAboutObsoleteDotNetCliToolReferences"
+          BeforeTargets="CollectPackageReferences"
+          Condition=" '$(SuppressObsoleteDotNetCliToolWarning)' != 'true' ">
+    <Warning Code="DOTNET1018"
+             Text="Using DotNetCliToolReference to reference '%(_ReferenceToObsoleteDotNetCliTool.Identity)' is obsolete and can be removed from this project. This tool is bundled by default in the .NET Core SDK."
```
Comment (nguerrera) [inline src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.ObsoleteReferences.targets]: This should be localized

File: src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.ObsoleteReferences.targets [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+  <ItemGroup>
+    <_ReferenceToObsoleteDotNetCliTool Include="@(DotNetCliToolReference)" />
+    <DotNetCliToolReference Remove="@(BundledDotNetCliToolReference)" />
+    <_ReferenceToObsoleteDotNetCliTool Remove="@(DotNetCliToolReference)" />
+  </ItemGroup>
+
+  <Target Name="_WarnAboutObsoleteDotNetCliToolReferences"
+          BeforeTargets="CollectPackageReferences"
+          Condition=" '$(SuppressObsoleteDotNetCliToolWarning)' != 'true' ">
+    <Warning Code="DOTNET1018"
```
Comment (nguerrera) [inline src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.ObsoleteReferences.targets]: How was this code chosen?

File: src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Sdk.targets [hunk 3]
```diff
@@ -394,7 +394,7 @@ Copyright (c) .NET Foundation. All rights reserved.
         <RunArguments Condition="'$(RunArguments)' == ''">$(StartArguments)</RunArguments>
       </PropertyGroup>
     </When>
-    
+
```
Comment (nguerrera) [inline src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Sdk.targets]: This file has tons of unrelated whitespace changes

**Review by nguerrera — CHANGES_REQUESTED** [review]
See above

File: src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.ObsoleteReferences.targets (hunk 1 above)
Comment (natemcmaster) [reply src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.ObsoleteReferences.targets]: Happy to. How?

File: src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.ObsoleteReferences.targets (hunk 2 above)
Comment (natemcmaster) [reply src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.ObsoleteReferences.targets]: Found the highest DOTNET code in https://github.com/dotnet/sdk/blob/master/src/Tasks/Common/Resources/Strings.resx and added one.
https://github.com/dotnet/sdk/blob/master/src/Tasks/Common/Resources/Strings.resx#L195

File: src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.ObsoleteReferences.targets (hunk 1 above)
Comment (nguerrera) [reply src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.ObsoleteReferences.targets]: We have NETSdkError task. We briefly had NETSdkWarning but I believe it was just deleted due to being unused. Resurrect it from history and use it.

File: src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.ObsoleteReferences.targets (hunk 2 above)
Comment (nguerrera) [reply src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.ObsoleteReferences.targets]: I think those are mostly left over from Nat defensively putting a bunch of messages from project.json era dotnet and nuget in strings.resx before v1 loc deadline just in case we needed them. We don't set Code elsewhere but probably should. I'd follow the current pattern here and we can file a bug to assign codes and use throughout.

File: src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.ObsoleteReferences.targets (hunk 1 above)
Comment (natemcmaster) [reply src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.ObsoleteReferences.targets]: Added NETSdkWarning.

File: src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.ObsoleteReferences.targets (hunk 2 above)
Comment (natemcmaster) [reply src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.ObsoleteReferences.targets]: Ok, opened this issue to follow up https://github.com/dotnet/sdk/issues/2065

File: src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Sdk.targets (hunk 3 above)
Comment (natemcmaster) [reply src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Sdk.targets]: Sorry for the noise. VS Code made me do it.

File: src/Tasks/Microsoft.NET.Build.Extensions.Tasks/msbuildExtensions/Microsoft/Microsoft.NET.Build.Extensions/Microsoft.NET.Build.Extensions.NETFramework.targets [hunk 4]
```diff
@@ -20,6 +20,7 @@ Copyright (c) .NET Foundation. All rights reserved.
 
   <UsingTask TaskName="GetDependsOnNETStandard" AssemblyFile="$(MicrosoftNETBuildExtensionsTasksAssembly)" />
   <UsingTask TaskName="NETBuildExtensionsError" AssemblyFile="$(MicrosoftNETBuildExtensionsTasksAssembly)" />
+  <UsingTask TaskName="NETBuildExtensionsWarning" AssemblyFile="$(MicrosoftNETBuildExtensionsTasksAssembly)" />
```
Comment (nguerrera) [inline src/Tasks/Microsoft.NET.Build.Extensions.Tasks/msbuildExtensions/Microsoft/Microsoft.NET.Build.Extensions/Microsoft.NET.Build.Extensions.NETFramework.targets]: We don't have any extensions warnings yet so this is unnecessary task load. We can wait for first use to add.

**Comment by natemcmaster** [comment]
By the way, if I want to get this into preview2, should I change the branch this PR targets?

File: src/Tasks/Microsoft.NET.Build.Extensions.Tasks/msbuildExtensions/Microsoft/Microsoft.NET.Build.Extensions/Microsoft.NET.Build.Extensions.NETFramework.targets (hunk 4 above)
Comment (nguerrera) [reply src/Tasks/Microsoft.NET.Build.Extensions.Tasks/msbuildExtensions/Microsoft/Microsoft.NET.Build.Extensions/Microsoft.NET.Build.Extensions.NETFramework.targets]: Actually load is clearly lazy extensions error is busted now due to build refactoring breaking ifdef, so this is ok.

**Review by nguerrera — APPROVED** [review]

**Comment by natemcmaster** [comment]
Ok. I'm going to retarget this PR to release/2.1.3xx to make sure this gets in for preview2.

**Comment by natemcmaster** [comment]
@dotnet-bot test Ubuntu14.04 Debug
@dotnet-bot test Ubuntu16.04 Release
@dotnet-bot test OSX10.12 Release
@dotnet-bot test Windows_NT Release

**Comment by natemcmaster** [comment]
@dotnet-bot test Windows_NT Release
@dotnet-bot test OSX10.12 Release

**Comment by livarcocc** [comment]
One last attempt.

@dotnet-bot test Windows_NT Release
@dotnet-bot test OSX10.12 Release

**Comment by livarcocc** [comment]
@johnbeisner is looking at flaky SDK tests for prod con. I think we are just hitting evidence of that in regular CI as well.

**Comment by natemcmaster** [comment]
@dotnet-bot test OSX10.12 Release  please

**Comment by natemcmaster** [comment]
CI pr tests finally passed ✅  :D
